pkg/ns/api.go

```diff
@@ -241,87 +241,132 @@ func (p *Provider) CreateFilesystem(params CreateFilesystemParams) error {
+		return err
 	}
 
-	uri := p.RestClient.BuildURI(
-		fmt.Sprintf("/storage/filesystems/%s", url.PathEscape(path)),
-		map[string]string{
-			"force":     "true",
-			"snapshots": destroySnapshotsString,
-		},
-	)
-
-	return p.sendRequest(http.MethodDelete, uri, nil)
-}
-
-// DestroyFilesystemWithClones destroys filesystem that has or may have clones
-// Method promotes filesystem's most recent snapshots and then delete the filesystem
-func (p *Provider) DestroyFilesystemWithClones(path string, destroySnapshots bool) (err error) {
-	l := p.Log.WithField("func", "DestroyFilesystemWithClones()")
+	// If here then filesystem deletion request has failed because
+	// the filesystem has dependent clones (EEXIST error code), trying
+	// to promote the most recent clone to make the filesystem independent:
 
 	maxAttemptCount := 3
+	var mostRecentError error
 
 	for i := 0; i < maxAttemptCount; i++ {
-		err = p.DestroyFilesystem(path, destroySnapshots)
-		if err == nil {
-			return nil
-		} else if IsNotExistNefError(err) {
-			return err
-		} else if !IsAlreadyExistNefError(err) { // EEXIST code there means the filesystem has dependent clones
-			continue
-		}
-
-		// if here then filesystem has dependent clones (EEXIST error on deletion),
-		// need promote the most recent clone to make the filesystem independent
+		mostRecentError = nil
 
 		snapshots, err := p.GetSnapshots(path, true)
 		if err != nil {
-			l.Warnf("filesystem '%s' has clones, but snapshots list was failed: %s (trying again...)", path, err)
-			continue
+			mostRecentError = fmt.Errorf("failed to get snapshot list: %s", err)
+			break
 		}
 
 		var maxCreationTxg int
 		var mostRecentClone string
 		for _, s := range snapshots {
-			// to get "clones" and "creationTxg" fields that are not presented int the list response
+			// to get "clones" and "creationTxg" fields that are not presented in the list response
 			snapshot, err := p.GetSnapshot(s.Path)
 			if err != nil {
-				l.Warnf(
-					"filesystem '%s' has clones, but getting snapshot '%s' was failed: %s (trying again...)",
-					path,
+				mostRecentError = fmt.Errorf("failed to get '%s' snapshost's info: %s", s.Path, err)
+				break
+			}
+			creationTxg, err := strconv.Atoi(snapshot.CreationTxg)
+			if err != nil {
+				mostRecentError = fmt.Errorf(
+					"snapshot '%s': failed to convert 'creationTxg' value '%s' to integer: %s",
 					s.Path,
+					snapshot.CreationTxg,
 					err,
 				)
-				continue
-			}
-			creationTxg, err := strconv.Atoi(snapshot.CreationTxg)
-			if len(snapshot.Clones) > 0 && creationTxg > maxCreationTxg {
+				break
+			} else if len(snapshot.Clones) > 0 && creationTxg > maxCreationTxg {
 				mostRecentClone = snapshot.Clones[0]
+				maxCreationTxg = creationTxg
 			}
 		}
-
-		if mostRecentClone == "" {
-			l.Warnf("filesystem '%s' is supposted to has clones, but it doesn't, trying again...", path)
+		if mostRecentError != nil {
```

Comment:
It seems this case is impossible

Reply:
"mostRecentError" could be set on https://github.com/Nexenta/go-nexentastor/commit/7cf5356ec827ec8ff566d436882145f4866005ef#diff-3d7d746e8181ca5cf7baf0ff3ef116e3R307 and `break` after it will lead to this condition. Right?